cmd/govim/highlight.go

```diff
@@ -86,18 +105,116 @@ func (v *vimstate) redefineHighlights(diags []types.Diagnostic, force bool) erro
 		v.BatchChannelCall("prop_add",
 			d.Range.Start.Line(),
 			d.Range.Start.Col(),
-			struct {
-				Type    string `json:"type"`
-				EndLine int    `json:"end_lnum"`
-				EndCol  int    `json:"end_col"` // column just after the text
-				BufNr   int    `json:"bufnr"`
-			}{string(hi), d.Range.End.Line(), d.Range.End.Col(), d.Buf})
+			propAddDict{string(hi), types.DiagnosticTextPropID, d.Range.End.Line(), d.Range.End.Col(), d.Buf},
+		)
+	}
+
+	v.BatchEnd()
+	return nil
+}
+
+func (v *vimstate) updateReferenceHighlight(refresh bool) error {
+	if v.config.HighlightReferences == nil || !*v.config.HighlightReferences {
+		return nil
+	}
+	b, pos, err := v.cursorPos()
+	if err != nil {
+		return fmt.Errorf("failed to get current position: %v", err)
+	}
+
+	// refresh indicates if govim should call DocumentHighlight to refresh
+	// ranges from gopls since we want to refresh when the user goes idle,
+	// and remove highlights as soon as the user is busy. To prevent
+	// flickering we keep track of the current highlight range and avoid
+	// removing text properties if the cursor is still within the range.
+	if !refresh {
+		if v.currentReference != nil && pos.IsWithin(*v.currentReference) {
+			return nil
+		}
+		v.currentReference = nil
+		v.removeTextProps(types.ReferencesTextPropID)
+		return nil
+	}
+
+	// Cancel any ongoing requests to make sure that we only process the
+	// latest response.
+	if v.cancelDocHighlight != nil {
+		v.cancelDocHighlight()
+		v.cancelDocHighlight = nil
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	v.cancelDocHighlight = cancel
+
+	v.tomb.Go(func() error {
+		v.redefineReferenceHighlight(ctx, b, pos)
+		cancel()
+		v.cancelDocHighlight = nil
+		return nil
+	})
+
+	return nil
+}
+
+func (v *vimstate) redefineReferenceHighlight(ctx context.Context, b *types.Buffer, cursorPos types.Point) {
+	res, err := v.server.DocumentHighlight(ctx,
+		&protocol.DocumentHighlightParams{
+			TextDocumentPositionParams: protocol.TextDocumentPositionParams{
+				TextDocument: protocol.TextDocumentIdentifier{
+					URI: string(b.URI()),
+				},
+				Position: cursorPos.ToPosition(),
+			},
+		},
+	)
+
+	if err == context.Canceled {
+		// A newer request canceled this one, so we shouldn't process the results here
+		return
+	} else if err != nil {
+		v.Logf("documentHighlight call failed: %v", err)
+		return
+	}
+
+	v.govimplugin.Schedule(func(govim.Govim) error {
+		return v.handleDocumentHighlight(b, cursorPos, res)
```

Comment:
Within this scheduled func we probably want another

```
select {
case <-ctx.Done():
    return
default:
}
```

Reply:
I don't think we should. Since I cancel the context to free up resources after the request to gopls, and the schedule execution might be scheduled far later, it might skip a valid execution.

I think that is what happened in the failed `gvim` tests (flakes when I had the select case). From the logs the only difference was a `Hover` request in the `gvim` logs, something I guess took enough time for us to hit this select-case return.

Reply:
Hmm, not sure. The reason I think you want to cancel here is that it's your last chance to do so before applying a potentially bad highlight. Because if you don't check the cancel here and the highlight you are about to apply has been cancelled, then it will look wrong until the the "correct" call returns.

Also I think you need to be re-checking whether the user is busy or not in here. Because if they are (which is the most likely cause of the current highlight having been cancelled) you definitely don't want to do anything.

Reply:
Yeah, I see the reasoning. I re-added the `ctx.Done()` check here, and added a busy check. I have a gut feeling that the busy check might cause unexpected behaviour, but can't seem to come up with such case.